clang/test/CodeGenHLSL/builtins/dot2add.hlsl

```diff
@@ -0,0 +1,78 @@
+float test_float_arg1_type(float2 p1, half2 p2, float p3) {
+  // CHECK-SPIRV:  %conv = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %1 to <2 x half>
+  // CHECK-SPIRV:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %2, float {{.*}} %3) #3 {{.*}}
+  // CHECK-DXIL:  %conv = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %0 to <2 x half>
+  // CHECK-DXIL:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %1, float {{.*}} %2) #2
+  // CHECK:  ret float %[[RES]]
+  return dot2add(p1, p2, p3);
+}
+
+// CHECK-LABEL: define {{.*}}test_double_arg3_type
+float test_double_arg3_type(half2 p1, half2 p2, double p3) {
+  // CHECK-SPIRV:  %conv = fptrunc reassoc nnan ninf nsz arcp afn double %3 to float
+  // CHECK-SPIRV:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %1, <2 x half> {{.*}} %2, float {{.*}} %conv) #3 {{.*}}
+  // CHECK-DXIL:  %conv = fptrunc reassoc nnan ninf nsz arcp afn double %2 to float
+  // CHECK-DXIL:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %0, <2 x half> {{.*}} %1, float {{.*}} %conv) #2
+  // CHECK:  ret float %[[RES]]
+  return dot2add(p1, p2, p3);
+}
+
+// CHECK-LABEL: define {{.*}}test_float_arg1_arg2_type
+float test_float_arg1_arg2_type(float2 p1, float2 p2, float p3) {
+  // CHECK-SPIRV:  %conv = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %1 to <2 x half>
+  // CHECK-SPIRV:  %conv1 = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %2 to <2 x half>
+  // CHECK-SPIRV:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %conv1, float {{.*}} %3) #3 {{.*}}
+  // CHECK-DXIL:  %conv = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %0 to <2 x half>
+  // CHECK-DXIL:  %conv1 = fptrunc reassoc nnan ninf nsz arcp afn <2 x float> %1 to <2 x half>
+  // CHECK-DXIL:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %conv1, float {{.*}} %2) #2
+  // CHECK:  ret float %[[RES]]
+  return dot2add(p1, p2, p3);
+}
+
+// CHECK-LABEL: define {{.*}}test_int16_arg1_arg2_type
+float test_int16_arg1_arg2_type(int16_t2 p1, int16_t2 p2, float p3) {
+  // CHECK-SPIRV:  %conv = sitofp <2 x i16> %1 to <2 x half>
+  // CHECK-SPIRV:  %conv1 = sitofp <2 x i16> %2 to <2 x half>
+  // CHECK-SPIRV:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %conv1, float {{.*}} %3) #3 {{.*}}
+  // CHECK-DXIL:  %conv = sitofp <2 x i16> %0 to <2 x half>
+  // CHECK-DXIL:  %conv1 = sitofp <2 x i16> %1 to <2 x half>
+  // CHECK-DXIL:  %[[RES:.*]] = call {{.*}} float @_ZN4hlsl7dot2addEDv2_DhS0_f(<2 x half> {{.*}} %conv, <2 x half> {{.*}} %conv1, float {{.*}} %2) #2
+  // CHECK:  ret float %[[RES]]
+  return dot2add(p1, p2, p3);
+}
+
+// CHECK-LABEL: define {{.*}}dot2add_impl

```

Comment:
I suppose this is one way to test the behavior and would let you keep `-disable-llvm-passes`. We typically would do this test per test case to account for the changing intrinsic definitions, but in your case only one intrinisc defintion is valid. I'll let others way in on this. I'm conflicted because I prefer to see checks associated with code written in the test not ones imported from our hlsl header. But I also understand that for dot2add we would be repeating these same checks verbatim for each test case.

Reply:
> I'll let others weigh in on this

I suggested that Sumit do it this way but that was mainly because I forgot about inlining. I do think enabling the passes risks testing the wrong thing/hiding an unintentional output but I also agree that it would be better to directly test that the opcode of interest is emitted

> But I also understand that for dot2add we would be repeating these same checks verbatim for each test case.
imo its okay to repeat a lot in test so this isn't a strong argument to keep the current thing